refactor LieTrace with BlockTriangular API

or4nge19 opened the pull request:
removed upperTriangular API to leverage BlockTriangular (recommended by Eric Wieser)

or4nge19 committed:
refactor LieTrace with BlockTriangular API

eric-wieser commented on PhysLean/Mathematics/DataStructures/Matrix/LieTrace.lean:
```diff
@@ -13,122 +14,55 @@ import PhysLean.Mathematics.SchurTriangulation
+  letI : CompleteSpace 𝕂 := by infer_instance
+  induction' s using Finset.induction with a s ha ih
+  · simp [NormedSpace.exp_zero]
+  · rw [Finset.prod_insert ha, Finset.sum_insert ha, NormedSpace.exp_add, ih]
```
This is `NormedSpace.exp_sum`

eric-wieser commented on PhysLean/Mathematics/DataStructures/Matrix/LieTrace.lean:
```diff
@@ -141,136 +75,123 @@ lemma matrix_tsum_apply
+
+/-- Powers of block triangular matrices are block triangular. -/
+lemma blockTriangular_pow {A : Matrix m m 𝕂} (hA : BlockTriangular A id) (k : ℕ) :
+    BlockTriangular (A ^ k) id := by
```
You should call this `BlockTriangular.pow` to make dot notation work like `mul`.

eric-wieser commented on PhysLean/Mathematics/DataStructures/Matrix/LieTrace.lean:
```diff
@@ -141,136 +75,123 @@ lemma matrix_tsum_apply
-    diag_exp_of_isUpperTriangular hA
-  erw [← Finset.prod_exp_eq_exp_sum Finset.univ]
-  exact congrArg Finset.univ.prod h_diag_exp
+  have h_det_U : det (U : Matrix m m 𝕂) * det (star (U : Matrix m m 𝕂)) = (1 : 𝕂) := by
```
`convert Matrix.det_conj U A` should shorten this further.

or4nge19 committed:
fix comment 1,2

I will try to address comment 3 tomorrow. Thanks!

or4nge19 committed:
removed unneccessary lemma

or4nge19 committed:
fix lint

or4nge19 committed:
golf det_unitary_conj

I haven't been able to use convert Matrix.det_conj U A (Lean seems to accept convert Matrix.det_conj _ A but can't get it work :( )

jstoobysmith reviewed (approved):
Approved - feel free to merge if you are ready. Happy to look at it again if you are not ready yet.
